File: Graficación por computadora (OpenGL)/Cliente/User11/practica2.py
from OpenGL.GLUT import *
from OpenGL.GLU import *
from OpenGL.GL import *
from math import *
import sys

retardo = 1 #retardo timer
width = 0
height = 0
#coordenadas centro del circulo mayor
centro_x = 300
centro_y = 300
#angulo del radio mayor
angulo1 = 0
#angulo del radio menor
angulo2 = 0

#rotacion del triangulo de parabolas
ang1 = 0

#coordenadas de las posiciones del triangulo de parabolas
coords = []

#indica la pausa del gif
pause = False

#indica el reinicio del gif
reinicio = False

cerrar = False

vista1 = True
vista2 = False

menu = False

def writeBitmapString(font, string):
##esta funcion imprime texto en pantalla
   for c in string:
       glutBitmapCharacter(font, ord(c))

#dibuja un circulo de acuerdo a las coordenadas y radio
def circulo(radio, X, Y):
	glColor3f(1,1,1)
	glBegin(GL_LINE_STRIP)
	for ang in range(360):
		x = radio * cos(radians(ang)) + X
		y = radio * sin(radians(ang)) + Y
		glVertex2f(x, y)
	glEnd()


def drawScene():

	global vertices, recorrido, vertice1, coordenadas1, coordenadas2, coordenadas3

	glClearColor(0,0,0,0)
	glClear(GL_COLOR_BUFFER_BIT)

	
	#Circulo mayor
	if vista1:
		circulo(300, centro_x, centro_y)
	#-------radio mayor------
	glColor3f(.5, 1, .5)
	glBegin(GL_LINES)	
	extremo_x = 225 * cos(radians(angulo1)) + centro_x
	extremo_y = 225 * sin(radians(angulo1)) + centro_y
	if vista1:
		glVertex2f(centro_x, centro_y)
		glVertex2f(extremo_x, extremo_y)
	glEnd()
	
	#Circulo mediano
	if vista1:
		circulo(75, extremo_x, extremo_y)

	#-------radio menor------
	glColor3f(1, .5, .5)
	glBegin(GL_LINES)	
	extremo_x2 = 25 * cos(radians(angulo2)) + extremo_x
	extremo_y2 = 25 * sin(radians(angulo2)) + extremo_y
	#coordenadas.append([extremo_x2, extremo_y2])
	if vista1:
		glVertex2f(extremo_x, extremo_y)
		glVertex2f(extremo_x2, extremo_y2)
	glEnd()

	#Circulo chico
	if vista1:
		circulo(50, extremo_x2, extremo_y2)

	r = 145
	r2 = 125

	##-------TRIANGULO DE PARABOLAS DENTRO DEL CIRCULO MEDIANO--------------

	if vista2:
		glPushMatrix()
		glTranslatef(extremo_x,extremo_y,0)
		glRotatef(ang1,0,0,1)
		#parabola1
		_x = r * cos(radians(60))
		_y = r * sin(radians(60))
		glColor3f(0,0,1)
		glBegin(GL_LINE_STRIP)
		for i in range(210,271):
			__x = r2 * cos(radians(i)) + _x
			__y = r2 * sin(radians(i)) + _y
			glVertex2f(__x, __y)
		glEnd()
		#parabola2
		_x = r * cos(radians(180))
		_y = r * sin(radians(180))
		glColor3f(0,0,1)
		glBegin(GL_LINE_STRIP)
		for i in range(330,391):
			__x = r2 * cos(radians(i)) + _x
			__y = r2 * sin(radians(i)) + _y
			glVertex2f(__x, __y)
		glEnd()
		#parabola3
		_x = r * cos(radians(300))
		_y = r * sin(radians(300))
		glColor3f(0,0,1)
		glBegin(GL_LINE_STRIP)
		for i in range(90,151):
			__x = r2 * cos(radians(i)) + _x
			__y = r2 * sin(radians(i)) + _y
			glVertex2f(__x, __y)
		glEnd()
		

		glPopMatrix()

		#se guardan las posiciones actuales del triangulo de parabolas
		coords.append([extremo_x, extremo_y, ang1])

		#se dibuja el recorrido de las parabolas con ayuda de las pociciones respaldadas en coords
		glColor3f(0,1,0)
		n = 0
		for c in coords:
			if n % 3 == 0:
				glPushMatrix()
				glTranslatef(c[0],c[1],0)
				glRotatef(c[2],0,0,1)
				#parabola1
				_x = r * cos(radians(60))
				_y = r * sin(radians(60))
				glColor3f(0,0,1)
				glBegin(GL_LINE_STRIP)
				for i in range(210,271):
					__x = r2 * cos(radians(i)) + _x
					__y = r2 * sin(radians(i)) + _y
					glVertex2f(__x, __y)
				glEnd()
				#parabola2
				_x = r * cos(radians(180))
				_y = r * sin(radians(180))
				glColor3f(0,0,1)
				glBegin(GL_LINE_STRIP)
				for i in range(330,391):
					__x = r2 * cos(radians(i)) + _x
					__y = r2 * sin(radians(i)) + _y
					glVertex2f(__x, __y)
				glEnd()
				#parabola3
				_x = r * cos(radians(300))
				_y = r * sin(radians(300))
				glColor3f(0,0,1)
				glBegin(GL_LINE_STRIP)
				for i in range(90,151):
					__x = r2 * cos(radians(i)) + _x
					__y = r2 * sin(radians(i)) + _y
					glVertex2f(__x, __y)
				glEnd()
				glPopMatrix()
			n+=1

	##--------------------------------------------

	if menu:
		#botones pausar
		glColor3f(1,1,1)
		glBegin(GL_POLYGON)
		glVertex2f(300, 300)
		glVertex2f(300, 320)
		glVertex2f(380, 320)
		glVertex2f(380, 300)
		glEnd()
		glColor3f(0,0,0)
		glRasterPos2f(310,305)
		writeBitmapString(GLUT_BITMAP_HELVETICA_12,"Pausar")
		#botones cerrar
		glColor3f(1,1,1)
		glBegin(GL_POLYGON)
		glVertex2f(300, 275)
		glVertex2f(300, 295)
		glVertex2f(380, 295)
		glVertex2f(380, 275)
		glEnd()
		glColor3f(0,0,0)
		glRasterPos2f(310,280)
		writeBitmapString(GLUT_BITMAP_HELVETICA_12,"Vista 1")
		#botones cerrar
		glColor3f(1,1,1)
		glBegin(GL_POLYGON)
		glVertex2f(300, 250)
		glVertex2f(300, 270)
		glVertex2f(380, 270)
		glVertex2f(380, 250)
		glEnd()
		glColor3f(0,0,0)
		glRasterPos2f(310,255)
		writeBitmapString(GLUT_BITMAP_HELVETICA_12,"Vista 2")
		#botones cerrar
		glColor3f(1,1,1)
		glBegin(GL_POLYGON)
		glVertex2f(300, 225)
		glVertex2f(300, 245)
		glVertex2f(380, 245)
		glVertex2f(380, 225)
		glEnd()
		glColor3f(0,0,0)
		glRasterPos2f(310,230)
		writeBitmapString(GLUT_BITMAP_HELVETICA_12,"Cerrar")

	glColor3f(1,1,1)
	glRasterPos2f(10,10)
	writeBitmapString(GLUT_BITMAP_HELVETICA_12,"Clic Derecho: Menu")
	
	glFlush()

def timer(value):
	global angulo1, angulo2, ang1, coords, pause

	#si el gif no esta pausado se cumple lo siguiente
	if pause == False:

		#se incrementan los angulos de rotacion de los radios
		angulo1 -= 2
		angulo2 += 22

		if angulo1 % 360 == 0:
			coords = []

		ang1 += 6

	#los datos se reinician a los valores predeterminados
	if reinicio:
		angulo1 = 0
		angulo2 = 0
		coords = []
		ang1 = 0
		pause = False

	#pausa = False

	glutPostRedisplay()
	glutTimerFunc(retardo ,timer, 1)

def mouseControl(button, state, x, y): 
	global pause, reinicio , menu, cerrar, vista1, vista2
	y = height - y

	##se da clic izquierdo para pausar
	if button == GLUT_LEFT_BUTTON and state == GLUT_DOWN:
		#coordenadas del boton de pause
		if x >= 300 and x <= 380 and y >= 300 and y <= 320:
			pause = not pause
		#coordenadas del boton de vista1
		if x >= 300 and x <= 380 and y >= 275 and y <= 295:
			vista1 = not vista1
		#coordenadas del boton de vista2
		if x >= 300 and x <= 380 and y >= 250 and y <= 270:
			vista2 = not vista2
		#coordenadas del boton de cerrar
		if x >= 300 and x <= 380 and y >= 225 and y <= 245:
			exit(0)

	##se da clic derecho para reinicio
	if button == GLUT_RIGHT_BUTTON and state == GLUT_DOWN:
		menu = not menu

	glutPostRedisplay()

def mouseMotion(x, y):
	y = height - y
	glutPostRedisplay()


def resize(w, h):
	global height, width
	glViewport (0, 0, w, h)
	glMatrixMode(GL_PROJECTION)
	glLoadIdentity()

	glOrtho(0.0, w, 0.0, h, -1.0, 1.0)

	width = w
	height = h 

	glMatrixMode(GL_MODELVIEW)
	glLoadIdentity()

def keyInput(key, x, y):
	glutPostRedisplay()

def main():
	glutInit(sys.argv)
	glutInitDisplayMode(GLUT_SINGLE | GLUT_RGBA)
	glutInitWindowSize(600, 600)
	glutInitWindowPosition(100, 100) 
	glutCreateWindow("Practica 2")
	glutDisplayFunc(drawScene)
	glutReshapeFunc(resize)
	glutKeyboardFunc(keyInput)
	glutMouseFunc(mouseControl)
	glutMotionFunc(mouseMotion)
	glutTimerFunc(retardo,timer, 1)
	glutMainLoop()
main()

File: 1. Ingeniería en tecnologías de la información/Graficación por computadora (OpenGL)/Cliente/practica26/practica2.cpp
#include <stdlib.h>
#include <iostream>
#include <math.h>
#include <stdio.h>
#include <GL/glut.h>
#include <sstream>
#include <vector>
#define SIZE 500
#define PI 3.141592654

using namespace std;

float width, height;

//coordenadas del origen del plano cartesiano
static float origenX = 80;
static float origenY = 130;

int vertices = 3;//numero de vertices inicial de la figura

vector <float> coordenadas[2];//vector que guarda las coordenadas de los vertices
vector <bool> vertice;//valor booleano de cada vertice que verifica si es presionado

bool angles = false;//activador de vista de los angulos
bool side = false;//activador de vista del valor del lado de la figura
bool diagonales = false;//activador de vista de las diagonales entre vertices
bool coords = false;//activador de vista de las coordenadas de cada vertice
bool guides = false;//activador de vista de los ejes fque forman cada vertice

void reshape(int w, int h)
{
    glViewport (0, 0, w, h);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();

    glOrtho(0.0, w, 0.0, h, -1.0, 1.0);

    width = w;
    height = h;

    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
}

void writeBitmapString(void *font, string str)
{
	int i;
   	for(i = 0; i < str.size(); i++)
   		glutBitmapCharacter(font, str[i]);
}

void display(void)
{
    glClearColor(1,1,1,0);
    glClear(GL_COLOR_BUFFER_BIT);

    //----------graficacion del plano cartesiano------------

    //--ejes
    glBegin(GL_LINES);
    //eje Y
    glColor3f(1,0,0);
    glVertex2f(origenX, 0);
    glVertex2f(origenX, 500);
    //eje X
    glColor3f(0,0,1);
    glVertex2f(0, origenY);
    glVertex2f(900, origenY);
    glEnd();


    glPointSize(5);

    //puntos del eje Y (Rojo)
    int separacion = 0;
    for(int i = 0; i < 7; i++){
    	stringstream ss; ss<<separacion<<endl;
   		string newStr = ss.str();
   		glBegin(GL_POINTS);
    	glColor3f(1,0,0);
    	glVertex2f(origenX, origenY + separacion);
    	glEnd();

    	//se dibujan los numeros del eje Y
    	glColor3f(1,0,0);
    	glRasterPos2f(origenX - 25, origenY + separacion);
   		writeBitmapString(GLUT_BITMAP_HELVETICA_12, newStr);

   		separacion += 50; //separacion entre puntos
    }

    //puntos del eje X (Azul)
    separacion = 0;
    for(int i = 0; i < 13; i++){
    	stringstream ss; ss<<separacion<<endl;
   		string newStr = ss.str();
    	glBegin(GL_POINTS);
    	glColor3f(0,0,1);
    	glVertex2f(origenX + separacion, origenY);
    	glEnd();

    	//se dibujan los numeros del exe X
    	glColor3f(0,0,1);
    	glRasterPos2f(origenX + separacion, origenY - 15);
   		writeBitmapString(GLUT_BITMAP_HELVETICA_12, newStr);

    	separacion += 50; //separacion entre puntos
    }

    //graficar vertices de la figura inicial
    glPointSize(10);
    glBegin(GL_POINTS);
    glColor3f(0,0,0);
    for(int i = 0; i < vertices; i++){
    	glVertex2f(coordenadas[0][i], coordenadas[1][i]);
    }
    glEnd();

    //muestra las diagonales que se forman entre todos los vertices de las figuras
    if(diagonales){
    	//se recorre cada vertice
    	for(int i = 0; i < vertices; i ++){
	    	glBegin(GL_LINES);
	    	glColor3f(.3,1,.3);
	    	//el vertice se une con los demas vertices
	    	for(int j = 0; j < vertices; j++){
	    		glVertex2f(coordenadas[0][i], coordenadas[1][i]);
	    		glVertex2f(coordenadas[0][j], coordenadas[1][j]);
	    	}
	    	glEnd();
	    }
    }

    //muestra las lineas de coordenadas de posicion del vertice
    if(guides){
    	//se recorre cada vertice
    	for(int i = 0; i < vertices; i ++){
	    	glBegin(GL_LINES);
	    	glColor3f(.8,.4,.8);
	    	//eje X vertice
	    	glVertex2f(origenX, coordenadas[1][i]);
	    	glVertex2f(coordenadas[0][i], coordenadas[1][i]);
	    	//eje Y vertice
	    	glVertex2f(coordenadas[0][i], coordenadas[1][i]);
	    	glVertex2f(coordenadas[0][i], origenY);
	    	glEnd();
	    }
    }

    //lineas (union de vertices)
    glBegin(GL_LINE_LOOP);
    glColor3f(0,0,0);
    for(int i = 0; i < vertices; i++){
    	glVertex2f(coordenadas[0][i], coordenadas[1][i]);
    }
    glEnd();

    float lado = sqrt(pow(coordenadas[0][0] - coordenadas[0][1], 2) + pow(coordenadas[1][0] - coordenadas[1][1], 2));

    //se dibujan los angulos inferiores

	for(int i = 0; i < vertices; i++){
    	float ang_inicio, ang_fin;
    	if(i == 0)
    		ang_inicio = atan2(coordenadas[1][vertices-1] - coordenadas[1][i], coordenadas[0][vertices-1] - coordenadas[0][i]) * 180/PI;
    	else
    		ang_inicio = atan2(coordenadas[1][i-1] - coordenadas[1][i], coordenadas[0][i-1] - coordenadas[0][i]) * 180/PI;

    	if(i == vertices-1)
    		ang_fin = atan2(coordenadas[1][0] - coordenadas[1][i], coordenadas[0][0] - coordenadas[0][i]) * 180/PI;
    	else
    		ang_fin = atan2(coordenadas[1][i+1] - coordenadas[1][i], coordenadas[0][i+1] - coordenadas[0][i]) * 180/PI;






    	if(ang_inicio < 0) ang_inicio += 360;
    	if(ang_fin < 0) ang_fin += 360;

    	//se dibuja el angulo de cada vertice
    	if(angles){
    		if(i == 0){
	    		//se muestra el valor del angulo inferior
	    		stringstream ss; ss<<ang_inicio-ang_fin<<endl;
	   			string newStr = ss.str();
	   			glColor3f(0,0,0);
		    	glRasterPos2f(coordenadas[0][0]-30, coordenadas[1][0]-50);
		   		writeBitmapString(GLUT_BITMAP_HELVETICA_12, "a= " + newStr);
	    	}
	    	glBegin(GL_LINE_STRIP);
	    	glColor3f(0,0,0);

	    	if(ang_fin > ang_inicio){
	    		for(int ang = ang_fin; ang < 360; ang++){
		    		float x = 20 * cos(ang * PI/180) + coordenadas[0][i];
		    		float y = 20 * sin(ang * PI/180) + coordenadas[1][i];
		    		glVertex2f(x, y);
		    	}
		    	for(int ang = 0; ang < ang_inicio; ang++){
		    		float x = 20 * cos(ang * PI/180) + coordenadas[0][i];
		    		float y = 20 * sin(ang * PI/180) + coordenadas[1][i];
		    		glVertex2f(x, y);
		    	}
	    	}
	    	else{
	    		for(int ang = ang_fin; ang < ang_inicio; ang++){
		    		float x = 20 * cos(ang * PI/180) + coordenadas[0][i];
		    		float y = 20 * sin(ang * PI/180) + coordenadas[1][i];
		    		glVertex2f(x, y);
		    	}
	    	}
	    	glEnd();
	    }

	    //muestra el valor de los lados
	    if(side){
	    	if(i == 0){
		   		//se muestra el valor del lado
	    		stringstream ss2; ss2<<lado<<endl;
	   			string newStr2 = ss2.str();
	   			glColor3f(0,0,0);
		    	glRasterPos2f(coordenadas[0][0]+60, coordenadas[1][0]-30);
		   		writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Side = " + newStr2);
	    	}
	    }

	    //muestra las coordenadas de cada vertice
	    if(coords){
	    	//se muestra el valor del lado
    		stringstream ss3, ss4;
    		ss3<<floor(coordenadas[0][i])-origenX<<endl;
    		ss4<<floor(coordenadas[1][i])-origenY<<endl;
   			string x_str = ss3.str();
   			string y_str = ss4.str();
   			glColor3f(1,0,.5);
	    	glRasterPos2f(coordenadas[0][i]+20, coordenadas[1][i]);
	   		writeBitmapString(GLUT_BITMAP_HELVETICA_12, "("+x_str+","+y_str+")");
	    }
    }

    glColor3f(0,0,0);
	glRasterPos2f(90, 80);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' + '  = incrementa vertices");
	glRasterPos2f(90, 70);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' - '  = decrementa vertices");
	glRasterPos2f(90, 60);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' a '  = angulos inferiores");
	glRasterPos2f(90, 50);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' s '  = medida del lado");
	glRasterPos2f(90, 40);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' d '  = diagonales");
	glRasterPos2f(90, 30);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' c '  = coordenadas");
	glRasterPos2f(90, 20);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' g '  = ejes vertices");
	glRasterPos2f(90, 90);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Tecla  ' r '  = restaurar");
	glRasterPos2f(90, 10);
	writeBitmapString(GLUT_BITMAP_HELVETICA_12, "Selecciona los vertices con el clic izquierdo para moverlos");



    glFlush();


}

void mouseControl(int button, int state, int x, int y){

    y = height - y;

    if(button == GLUT_LEFT_BUTTON && state == GLUT_DOWN){
        //se verifica que los vertices sean seleccionados con el mouse
    	for(int i = 0; i < vertices; i++){
    		if(x >= coordenadas[0][i]-5 && x <= coordenadas[0][i]+5 && y >= coordenadas[1][i]-5 && y <= coordenadas[1][i]+5){
    			vertice[i] = true;
    		}
    		else{
    			vertice[i] = false;
    		}
    	}

    }

    glutPostRedisplay();
}

void mouseMotion(int x, int y){

    y = height - y;

    //se mueven los vertices de la figura de acuerdo a las posiciones del mouse
    for(int i = 0; i < vertices; i++){
    	if(vertice[i]){
    		coordenadas[0][i] = x;
    		coordenadas[1][i] = y;
    	}
    }


    glutPostRedisplay();
}


void keyInput(unsigned char key, int x, int y)
{
	float distancia_angulo;
    float angulo_inicial = 70.0;
   switch (key)
   {
      case '+':
      	vertices++;
      	coordenadas[0].clear();
      	coordenadas[1].clear();
      	distancia_angulo = 360.0 / vertices;
      	for(int i = 0; i < vertices; i++){
	    	float x = 150 * cos(angulo_inicial * PI / 180) + 300;
	    	float y = 150 * sin(angulo_inicial * PI / 180) + 280;
	    	coordenadas[0].push_back(x);
	    	coordenadas[1].push_back(y);
	    	vertice.push_back(false);
	    	angulo_inicial += distancia_angulo;
	    }
        break;
      case '-':
      	if(vertices > 3){
      		vertices--;
      		coordenadas[0].clear();
      		coordenadas[1].clear();
      		distancia_angulo = 360 / vertices;
      		for(int i = 0; i < vertices; i++){
		    	float x = 150 * cos(angulo_inicial * PI / 180) + 300;
		    	float y = 150 * sin(angulo_inicial * PI / 180) + 280;
		    	coordenadas[0].push_back(x);
		    	coordenadas[1].push_back(y);
		    	vertice.push_back(false);
		    	angulo_inicial += distancia_angulo;
		    }
      	}
        break;
      case 'r':
      	//vertices++;
      	coordenadas[0].clear();
      	coordenadas[1].clear();
      	distancia_angulo = 360.0 / vertices;
      	for(int i = 0; i < vertices; i++){
	    	float x = 150 * cos(angulo_inicial * PI / 180) + 300;
	    	float y = 150 * sin(angulo_inicial * PI / 180) + 280;
	    	coordenadas[0].push_back(x);
	    	coordenadas[1].push_back(y);
	    	vertice.push_back(false);
	    	angulo_inicial += distancia_angulo;
	    }
        break;
        break;

      case 'a':
      	angles = !angles;
      	break;
      case 's':
      	side = !side;
      	break;
      case 'd':
      	diagonales = !diagonales;
      	break;
      case 'c':
      	coords = !coords;
      	break;
      case 'g':
      	guides = !guides;
      	break;
      default:
         break;
   }
   glutPostRedisplay();
}

// Funcion principal.
int main(int argc, char *argv[])
{
	float distancia_angulo = 360 / vertices;
    float angulo_inicial = 70.0;

	for(int i = 0; i < vertices; i++){
    	float x = 150 * cos(angulo_inicial * PI / 180) + 300;
    	float y = 150 * sin(angulo_inicial * PI / 180) + 280;
    	coordenadas[0].push_back(x);
    	coordenadas[1].push_back(y);
    	vertice.push_back(false);
    	angulo_inicial += distancia_angulo;
    }
    cout << "hola" << endl;
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGBA);
    glutInitWindowSize(900, 500);
    glutInitWindowPosition(10, 10);
    glutCreateWindow("practica2");
    glutDisplayFunc(display);
    glutKeyboardFunc(keyInput);
    glutReshapeFunc(reshape);
    glutMouseFunc(mouseControl);
	glutMotionFunc(mouseMotion);
    glutMainLoop();
    return 0;
}
